x/ecocredit/server/marketplace/sell.go

```diff
@@ -0,0 +1,89 @@
+package marketplace
+
+import (
+	"context"
+	"fmt"
+	"github.com/cosmos/cosmos-sdk/orm/types/ormerrors"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+	marketApi "github.com/regen-network/regen-ledger/api/regen/ecocredit/marketplace/v1"
+	"github.com/regen-network/regen-ledger/types/math"
+	"github.com/regen-network/regen-ledger/x/ecocredit"
+	marketplacev1 "github.com/regen-network/regen-ledger/x/ecocredit/marketplace"
+	"github.com/regen-network/regen-ledger/x/ecocredit/server"
+	"google.golang.org/protobuf/types/known/timestamppb"
+)
+
+// Sell creates new sell orders for credits
+func (k Keeper) Sell(ctx context.Context, req *marketplacev1.MsgSell) (*marketplacev1.MsgSellResponse, error) {
+	sdkCtx := sdk.UnwrapSDKContext(ctx)
+	ownerAcc, err := sdk.AccAddressFromBech32(req.Owner)
+	if err != nil {
+		return nil, err
+	}
+
+	sellOrderIds := make([]uint64, len(req.Orders))
+
+	for i, order := range req.Orders {
+		batch, err := k.coreStore.BatchInfoStore().GetByBatchDenom(ctx, order.BatchDenom)
+		if err != nil {
+			return nil, fmt.Errorf("batch denom %v: %v", order.BatchDenom, err)
+		}
+		ct, err := server.GetCreditTypeFromBatchDenom(ctx, k.coreStore, k.params, batch.BatchDenom)
+		if err != nil {
+			return nil, err
+		}
+		marketId, err := k.getOrCreateMarketId(ctx, ct.Abbreviation, order.AskPrice.Denom)
+		// verify expiration is in the future
+		if order.Expiration != nil && order.Expiration.Before(sdkCtx.BlockTime()) {
+			return nil, sdkerrors.ErrInvalidRequest.Wrapf("expiration must be in the future: %s", order.Expiration)
+		}
+		sellQty, err := math.NewDecFromString(order.Quantity)
+		if err != nil {
+			return nil, err
+		}
+		if err := assertHasBalance(ctx, k.coreStore, ownerAcc, batch.Id, sellQty); err != nil {
+			return nil, err
+		}
+		has, err := isDenomAllowed(ctx, k.stateStore, order.AskPrice.Denom)
+		if err != nil {
+			return nil, err
+		}
```

Comment:
This should be commented out for now and linked to #624, which we are waiting for v0.46 to implement. There is no way to add an ask denom at the moment because this requires the gov module updates.